nipoppy/workflows/pipeline.py

```diff
@@ -400,21 +404,99 @@ def run_main(self):
             analysis_level=self.pipeline_step_config.ANALYSIS_LEVEL,
         )
 
+        if self.hpc:
+            self.submit_hpc_job(participants_sessions)
+        else:
+            # Default behavior
+            for participant_id, session_id in participants_sessions:
+                self.n_total += 1
+                self.logger.info(
+                    f"Running for participant {participant_id}, session {session_id}"
+                )
+                try:
+                    self.run_single(participant_id, session_id)
+                    self.n_success += 1
+                except Exception as exception:
+                    self.return_code = ReturnCode.PARTIAL_SUCCESS
+                    self.logger.error(
+                        f"Error running {self.pipeline_name} {self.pipeline_version}"
+                        f" on participant {participant_id}, session {session_id}"
+                        f": {exception}"
+                    )
+
+    def submit_hpc_job(self, participants_sessions):
+        """Submit jobs to a HPC cluster for processing."""
+        self.logger.info("Running in HPC mode.")
+
+        hpc_templates_path = Path(f"{self.dpath_root}/code/hpc_templates/{self.hpc}")
+        qa = QueueAdapter(directory=str(hpc_templates_path))
+
+        # Generate the list of nipoppy commands as a single string for a shell array
+        job_array_commands = []
         for participant_id, session_id in participants_sessions:
-            self.n_total += 1
+            command = [
+                "nipoppy",
+                "run",
+                str(self.dpath_root),
+                "--pipeline",
+                self.pipeline_name,
+                "--pipeline-version",
+                self.pipeline_version,
+                "--pipeline-step",
+                self.pipeline_step,
+                "--participant-id",
+                participant_id,
+                "--session-id",
+                session_id,
+            ]
+            job_array_commands.append(shlex.join(command))
+        # Join the commands into a single string
+
+        job_array_commands_str = " ".join([f'"{cmd}"' for cmd in job_array_commands])
+        module_load = self.config.HPC_PREAMBLE
+
+        # Build the single command to submit as an array job
+        if self.hpc == "slurm":
+            command = (
+                f"bash -c '{module_load}; commands=({job_array_commands_str}); "
+                f'eval "${{commands[$SLURM_ARRAY_TASK_ID]}}"\''
+            )
+        elif self.hpc == "sge":
+            command = (
+                f"bash -c '{module_load}; commands=({job_array_commands_str}); "
+                f'eval "${{commands[$((SGE_TASK_ID-1))]}}"\''
+            )
+        else:
+            raise ValueError(
```

Comment:
This can also be tested

Reply:
I thought I'd already tested if this ValueError was raised with unsupported HPC types (you commented on it above)?